### PR TITLE
[SPARK-9698] [ML] Add RInteraction transformer for supporting R-style feature interactions

### PR DESCRIPTION
**Pull request by ericl**
This is a pre-req for supporting the ":" operator in the RFormula feature transformer.

Design doc from umbrella task: https://docs.google.com/document/d/10NZNSEurN2EdWM31uFYsgayIPfCFHiuIu3pCWrUmP_c/edit

@mengxr

### CODE REVIEW

**Comment by mengxr** [comment]
@ericl Shall we split this PR into two?

1. Add `Interaction` as a transformer (SPARK-9698).
2. Support feature interaction in RFormula.

After 1) is merged, people can start working on the Python API, without being blocked by 2).

**Comment by ericl** [comment]
@mengxr done, this PR now just has the RInteraction changes.

**Comment by dputler** [comment]
I'm not clear as to how the order operation is determined. Looking at the tests, in the case of a categorical interaction it appears that it is based on the order in which unique category values are encountered for a categorical variable. Specifically, for the numeric/categorical interaction, the last category encountered ("baz") provides the first values of the interaction values, and the first category encountered ("foo") provides the last values of the interaction. In contrast, for the interaction between two categorical variables, the column order is set by the first category of the second underlying categorical variable (the value zq) is primary in column ordering (with zq-bar being the first column), so order is used again, but it runs in opposite direction for the two variables. This structure will actually work fine for model training, however, things get more complicated for predicting new data with this model. The approach is basically the same approach as MS/Revolution uses in their Revo ScaleR package (i.e., the order of the categories depends on when they are first encountered in the data). However, this turns out to greatly complicate predicting new data with a Revo ScaleR model in practice. Open source R works by first determining all the category labels for each categorical variable, alphabetically sorts the unique label for each categorical variable, and then basis the new feature order on the alphabetical sort of category labels, so the order in which a category label is encountered does not matter. This turns out to make dealing with predicting new data with an existing model much easier. The cost is the data needs to be passed over twice, with the first determining the set of unique category labels.

**Comment by mengxr** [comment]
@dputler Under distributed setting, we need to make at least one pass to collect all categories. The ordering is not alphabetical but by frequency (https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala#L86). The most frequent category gets index `0`, and one-hot encoder by default drops the last category, i.e., the least frequent one.

**Comment by dputler** [comment]
That actually doesn't deal with the scoring issue. What happens when new data to be predicted from an existing model has a more frequent category in a categorical variable than was the case in the training data? What happens if this is included in a Spark Streaming scoring process when the batch size might be one? As before, the frequency base indexing works for estimation, but will cause heartburn in many cases when trying to predict new data with an existing model.

**Comment by ericl** [comment]
If if I understand correctly, the concern is that the category to index assignment when predicting data will be different from that used when fitting the model. This should be OK here since `StringIndexer` retains a mapping from category to indices, which is reused when calling predict() on the model later.

It is true that it would be nice to have a more predictable ordering (such as alphabetic) for some tasks like comparing coefficients between different models, but I think this is should be a feature of `StringIndexer` and is not really related to this PR.

**Comment by dputler** [comment]
Thanks Eric, I knew that was it really wasn't an issue related to this specific pull request, but one that exposed an issue that, from past experience, looked like could come up again. I'll move the discussion to more appropriate forum.

**Comment by mengxr** [comment]
Had an offline discussion with @ericl 
* split this PR into two
  1. an `Interaction` transformer that only handles ML columns (`Double` and `Vector` with attributes for categorial features)
  2. move the string column handling logic to `RFormula`

**Comment by ericl** [comment]
@mengxr I did the refactoring as suggested

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.ml.{Estimator, Model, Pipeline, PipelineModel, PipelineStage, Transformer}
+import org.apache.spark.mllib.linalg.{Vector, VectorUDT, Vectors}
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+/**
+ * :: Experimental ::
+ * Implements the transforms required for R-style feature interactions. This transformer takes in
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: remove `R-style` because this is a common feature transformation

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  /** @group setParam */
+  def setInputCols(values: Array[String]): this.type = set(inputCols, values)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  // optimistic schema; does not contain any ML attributes
+  override def transformSchema(schema: StructType): StructType = {
+    checkParams()
+    StructType(schema.fields :+ StructField($(outputCol), new VectorUDT, true))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: `true` -> `false`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        case _: NumericType | BooleanType => dataset(c).cast(DoubleType)
+      }
+    }
+    val attrs = generateAttrs($(inputCols).map(col => dataset.schema(col)))
+    dataset.select(
+      col("*"),
+      interactFunc(struct(args: _*)).as($(outputCol), attrs.toMetadata()))
+  }
+
+  private def getIterators(dataset: DataFrame): Array[EncodingIterator] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: add API doc

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    def getCardinality(attr: Attribute): Int = {
+      attr match {
+        case nominal: NominalAttribute =>
+          nominal.getNumValues.getOrElse(
+            throw new SparkException("Nominal fields must have attr numValues defined."))
+        case _ =>
+          0  // treated as numeric
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: This is used to calculate the output index of iteration features. Maybe we can change `cardinality` to `numFeatures`. Then the output is `numValues` for a nominal feature, 1 for a binary feature, and 1 for a numeric feature.

Btw, (not in this PR) it might be useful to store only the nonzeros because most features would be numeric features. If there are millions of features, this cardinality array becomes very large.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        case _: VectorUDT =>
+          val attrs = AttributeGroup.fromStructField(field).attributes.getOrElse(
+            throw new SparkException("Vector attributes must be defined for interaction."))
+          attrs.map(getCardinality).toArray
+      }
+      new EncodingIterator(cardinalities)
+    }.toArray
+  }
+
+  private def generateAttrs(schema: Seq[StructField]): AttributeGroup = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: * missing API doc
* `schema` -> `fields` (`schema` should be a `StructType`)

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    var attrs: Seq[Attribute] = Nil
+    schema.reverse.map { field =>
+      val attrIterator = field.dataType match {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: `attrIterator` is not an `Iterator`. rename it to `encodedAttrs`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+          attrs.map { tail =>
+            NumericAttribute.defaultAttr.withName(head.name.get + ":" + tail.name.get)
+          }
+        }
+      }
+    }
+    new AttributeGroup($(outputCol), attrs.toArray)
+  }
+
+  private def getAttributesAfterEncoding(
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: * missing doc
* it might be helpful to list the combinations in the API doc

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+      groupName: Option[String], attrs: Seq[Attribute]): Seq[Attribute] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: chop down args

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+            j => BinaryAttribute.defaultAttr.withName(format(i, nominal.name, Some(j.toString))))
+        }
+      case (a: Attribute, i) =>
+        Seq(NumericAttribute.defaultAttr.withName(format(i, a.name, None)))
+    }
+  }
+
+  override def copy(extra: ParamMap): Interaction = defaultCopy(extra)
+
+  private def checkParams(): Unit = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: should override `validateParams`

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    require(get(outputCol).isDefined, "Output col must be defined first.")
+    require($(inputCols).length > 0, "Input cols must have non-zero length.")
+    require($(inputCols).distinct.length == $(inputCols).length, "Input cols must be distinct.")
+  }
+}
+
+/**  
+ * An iterator over VectorUDT or Double that one-hot encodes nominal fields in the output.
+ *
+ * @param cardinalities An array defining the cardinality of each vector sub-field, or a single
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: It may be simpler to store the cumulative sum of `cardinalities` instead.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+ *                      value if the field is numeric. Fields with zero cardinality will not be
+ *                      one-hot encoded (output verbatim).
+ */
+// TODO(ekl) support drop-last option like OneHotEncoder does?
+private[ml] class EncodingIterator(cardinalities: Array[Int]) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: minor: This is not really an `Iterator`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+  /** The size of the output vector. */
+  val size = cardinalities.map(i => if (i > 0) i else 1).sum
+
+  /**
+   * @param value The row to iterate over, either a Double or Vector.
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: missing doc

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+   * @param f The callback to invoke on each non-zero (index, value) output pair.
+   */
+  def foreachActive(value: Any, f: (Int, Double) => Unit) = value match {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: `foreachActive` -> `foreachNonzero`

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+          d >= 0.0 && d == d.toInt && d < numOutputCols,
+          s"Values from column must be indices, but got $d.")
+        f(d.toInt, 1.0)
+      } else {
+        f(0, d)
+      }
+    case vec: Vector =>
+      assert(cardinalities.length == vec.size,
+        s"Vector column size was ${vec.size}, expected ${cardinalities.length}")
+      val dense = vec.toDense
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: This is too expensive for sparse vectors. We should call `vec.foreachActive`. Storing the cumulative number of features could help.

**Comment by ericl** [comment]
@mengxr made the requested changes. I found it simpler to keep `numFeatures` in combination with an array of offsets instead of just the cumulative count though.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import scala.collection.mutable.{ArrayBuffer, ArrayBuilder}
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: `ArrayBuffer` is not used.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+
+import org.apache.spark.SparkException
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.ml.attribute._
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.param.shared._
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.ml.{Estimator, Model, Pipeline, PipelineModel, PipelineStage, Transformer}
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: Only `Transformer` is used.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+private[ml] class FeatureEncoder(numFeatures: Array[Int]) {
+  assert(numFeatures.forall(_ > 0), "Features counts must all be positive.")
+
+  /** The size of the output vector. */
+  val outputSize = numFeatures.sum
+
+  /** Precomputed offsets for the location of each output feature. */
+  private val outputOffsets = {
+    val arr = new Array[Int](numFeatures.length)
+    for (i <- 1 until arr.length) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: minor: Use `while` instead of `for`. The latter is slow in Scala. Another option is using `foldLeft` (not as fast as `while` but saves some code)

**Comment by mengxr** [comment]
LGTM except minor comments.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala (hunk 15 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala (hunk 16 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala (hunk 17 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala]: Done

**Comment by mengxr** [comment]
Merged into master. Thanks!
